File: robot/Comm/Messages/Remote/RemoteMessageToRobot.cpp
#include "RemoteMessageToRobot.h"
#include "Comm/Messages/Remote/ToRobot/SetCameraMessage.h"
#include "Comm/Messages/Remote/ToRobot/StartRemoteControlMessage.h"
#include "Comm/Messages/Remote/ToRobot/StopRemoteControlMessage.h"
#include "Comm/Messages/Remote/ToRobot/SetConfigValueMessage.h"
#include "Comm/Messages/Remote/ToRobot/SetHeadStiffnessMessage.h"
#include "Comm/Messages/Remote/ToRobot/RestoreHeadStiffnessMessage.h"

#include <cstring>

RemoteMessageToRobot::RemoteMessageToRobot(int _type, unsigned int _size, void const * _data)
  : RemoteMessage(_type, _size, _data),
    type(unknown) {
  if ((_type >= 0) && (_type < RemoteMessageToRobot::maxTypes)) {
    type = (RemoteMessageToRobot::MessageType)_type;
  }
}

RemoteMessageToRobot::RemoteMessageToRobot(MessageType _type, unsigned int _size, void const * _data)
  : RemoteMessage((int)_type, _size, _data),
    type(_type) {
}

RemoteMessageToRobot::~RemoteMessageToRobot() {
}

RemoteMessageToRobot::MessageType RemoteMessageToRobot::getType() const {
  return type;
}

RemoteMessageToRobot const * RemoteMessageToRobot::create(int typeInt, unsigned int size, void const * data) {
  RemoteMessageToRobot::MessageType type = ((typeInt < 0) || (typeInt >= RemoteMessageToRobot::maxTypes)) ? RemoteMessageToRobot::unknown : (RemoteMessageToRobot::MessageType)typeInt;

  switch (type) {
  case RemoteMessageToRobot::startRemoteControl:
    return new StartRemoteControlMessage();
    break;
  case RemoteMessageToRobot::stopRemoteControl:
    return new StopRemoteControlMessage();
    break;
  case RemoteMessageToRobot::setCamera:
    return SetCameraMessage::create(size, data);
    break;
  case RemoteMessageToRobot::setConfigValue:
    return SetConfigValueMessage::create(size, data);
  case RemoteMessageToRobot::setHeadStiffness:
    return SetHeadStiffnessMessage::create(size, data);
  case RemoteMessageToRobot::restoreHeadStiffness:
    return new RestoreHeadStiffnessMessage();
  default:
    if (size > 0) {
      return new RemoteMessageToRobot(RemoteMessageToRobot::unknown, size, data);
    }
    else {
      return new RemoteMessageToRobot(RemoteMessageToRobot::unknown, 0, NULL);
    }
  }
}

File: tools/SharedJavaLibrary/src/network/message/RemoteMessageToRobot.java
package network.message;

public class RemoteMessageToRobot extends RemoteMessage {

  public enum MessageType {
    // This should match Comm/Messages/Remote/RemoteMessageToRobot.h
    UNKNOWN(0),
    START_REMOTE_CONTROL(1),
    STOP_REMOTE_CONTROL(2),
    SET_CAMERA(3),
    SET_CONFIG_VALUE(4),
    SET_HEAD_STIFFNESS(5),
    RESTORE_HEAD_STIFFNESS(6);
    
    MessageType(int type) {
      this.type = type;
    }
    
    public int getTypeID() {
      return type;
    }
    
    private int type;
    
    public static MessageType create(int type) {
      for (MessageType messageType : MessageType.values()) {
        if (messageType.type == type) {
          return messageType;
        }
      }
        
      return MessageType.UNKNOWN;
    }
  }
  
  public static final int MESSAGE_HEADER_SIZE = 5;

  public RemoteMessageToRobot(MessageType type, byte [] data) {
    super(type.getTypeID(), data);
    
    this.type = type;
  }
  
  public MessageType getType() {
    return type;
  }
  
  public byte [] convertToBytes() {
    int dataLength = 0;
    if (data != null) {
      dataLength = data.length;
    }
    
    byte [] bytes = new byte[dataLength + MESSAGE_HEADER_SIZE];
    
    // The first byte is the type
    bytes[0] = (byte)(type.getTypeID() & 0xFF);
    // The next 4 bytes are the data length
    if (!writeInt(bytes, 1, dataLength)) {
      System.err.println("Error writing data length.");
      return null;
    }
    
    if (dataLength > 0) {
      System.arraycopy(data, 0, bytes, MESSAGE_HEADER_SIZE, dataLength);
    }
    
    return bytes;
  }
  
  private MessageType type;
}
